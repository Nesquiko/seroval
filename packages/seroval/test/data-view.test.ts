import { describe, it, expect } from 'vitest';
import {
  deserialize,
  fromJSON,
  serialize,
  serializeAsync,
  toJSONAsync,
  toJSON,
  crossSerialize,
  crossSerializeAsync,
  crossSerializeStream,
} from '../src';

describe('DataView', () => {
  describe('serialize', () => {
    it('supports DataView', () => {
      const buffer = new ArrayBuffer(16);
      const example = new DataView(buffer, 0);
      example.setInt16(1, 42);
      const result = serialize(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<DataView>(result);
      expect(back).toBeInstanceOf(DataView);
      expect(back.getInt16(1)).toBe(example.getInt16(1));
    });
  });
  describe('serializeAsync', () => {
    it('supports DataView', async () => {
      const buffer = new ArrayBuffer(16);
      const example = new DataView(buffer, 0);
      example.setInt16(1, 42);
      const result = await serializeAsync(Promise.resolve(example));
      expect(result).toMatchSnapshot();
      const back = await deserialize<Promise<DataView>>(result);
      expect(back).toBeInstanceOf(DataView);
      expect(back.getInt16(1)).toBe(example.getInt16(1));
    });
  });
  describe('toJSON', () => {
    it('supports DataView', () => {
      const buffer = new ArrayBuffer(16);
      const example = new DataView(buffer, 0);
      example.setInt16(1, 42);
      const result = toJSON(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<DataView>(result);
      expect(back).toBeInstanceOf(DataView);
      expect(back.getInt16(1)).toBe(example.getInt16(1));
    });
  });
  describe('toJSONAsync', () => {
    it('supports DataView', async () => {
      const buffer = new ArrayBuffer(16);
      const example = new DataView(buffer, 0);
      example.setInt16(1, 42);
      const result = await toJSONAsync(Promise.resolve(example));
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = await fromJSON<Promise<DataView>>(result);
      expect(back).toBeInstanceOf(DataView);
      expect(back.getInt16(1)).toBe(example.getInt16(1));
    });
  });
  describe('crossSerialize', () => {
    it('supports DataView', () => {
      const buffer = new ArrayBuffer(16);
      const example = new DataView(buffer, 0);
      example.setInt16(1, 42);
      const result = crossSerialize(example);
      expect(result).toMatchSnapshot();
    });
  });
  describe('crossSerializeAsync', () => {
    it('supports DataView', async () => {
      const buffer = new ArrayBuffer(16);
      const example = new DataView(buffer, 0);
      example.setInt16(1, 42);
      const result = await crossSerializeAsync(Promise.resolve(example));
      expect(result).toMatchSnapshot();
    });
  });
  describe('crossSerializeStream', () => {
    it('supports boxed bigint', async () => new Promise<void>((done) => {
      const buffer = new ArrayBuffer(16);
      const example = new DataView(buffer, 0);
      example.setInt16(1, 42);
      crossSerializeStream(Promise.resolve(example), {
        onSerialize(data) {
          expect(data).toMatchSnapshot();
        },
        onDone() {
          done();
        },
      });
    }));
  });
});
