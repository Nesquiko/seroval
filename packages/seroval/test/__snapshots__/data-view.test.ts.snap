// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DataView > crossSerialize > scoped > supports DataView 1`] = `"($R=>$R[0]=new DataView($R[1]=new Uint8Array([0,0,42,0,0,0,0,0,0,0,0,0,0,0,0,0]).buffer,0,16))($R[\\"example\\"])"`;

exports[`DataView > crossSerialize > supports DataView 1`] = `"$R[0]=new DataView($R[1]=new Uint8Array([0,0,42,0,0,0,0,0,0,0,0,0,0,0,0,0]).buffer,0,16)"`;

exports[`DataView > crossSerializeAsync > scoped > supports DataView 1`] = `"($R=>$R[0]=Promise.resolve($R[1]=new DataView($R[2]=new Uint8Array([0,0,42,0,0,0,0,0,0,0,0,0,0,0,0,0]).buffer,0,16)))($R[\\"example\\"])"`;

exports[`DataView > crossSerializeAsync > supports DataView 1`] = `"$R[0]=Promise.resolve($R[1]=new DataView($R[2]=new Uint8Array([0,0,42,0,0,0,0,0,0,0,0,0,0,0,0,0]).buffer,0,16))"`;

exports[`DataView > crossSerializeStream > scoped > supports DataView 1`] = `"($R=>$R[0]=($R[1]=(s,f,p)=>((p=new Promise((a,b)=>{s=a,f=b})).s=s,p.f=f,p))())($R[\\"example\\"])"`;

exports[`DataView > crossSerializeStream > scoped > supports DataView 2`] = `"($R=>($R[4]=(p,d)=>{p.s(d),p.status=\\"success\\",p.value=d;delete p.s;delete p.f})($R[0],$R[2]=new DataView($R[3]=new Uint8Array([0,0,42,0,0,0,0,0,0,0,0,0,0,0,0,0]).buffer,0,16)))($R[\\"example\\"])"`;

exports[`DataView > crossSerializeStream > supports DataView 1`] = `"$R[0]=($R[1]=(s,f,p)=>((p=new Promise((a,b)=>{s=a,f=b})).s=s,p.f=f,p))()"`;

exports[`DataView > crossSerializeStream > supports DataView 2`] = `"($R[4]=(p,d)=>{p.s(d),p.status=\\"success\\",p.value=d;delete p.s;delete p.f})($R[0],$R[2]=new DataView($R[3]=new Uint8Array([0,0,42,0,0,0,0,0,0,0,0,0,0,0,0,0]).buffer,0,16))"`;

exports[`DataView > serialize > supports DataView 1`] = `"new DataView(new Uint8Array([0,0,42,0,0,0,0,0,0,0,0,0,0,0,0,0]).buffer,0,16)"`;

exports[`DataView > serializeAsync > supports DataView 1`] = `"Promise.resolve(new DataView(new Uint8Array([0,0,42,0,0,0,0,0,0,0,0,0,0,0,0,0]).buffer,0,16))"`;

exports[`DataView > toCrossJSON > supports DataView 1`] = `"{\\"t\\":22,\\"i\\":0,\\"l\\":16,\\"f\\":{\\"t\\":21,\\"i\\":1,\\"s\\":[0,0,42,0,0,0,0,0,0,0,0,0,0,0,0,0]},\\"b\\":0}"`;

exports[`DataView > toCrossJSONAsync > supports DataView 1`] = `"{\\"t\\":12,\\"i\\":0,\\"s\\":1,\\"f\\":{\\"t\\":22,\\"i\\":1,\\"l\\":16,\\"f\\":{\\"t\\":21,\\"i\\":2,\\"s\\":[0,0,42,0,0,0,0,0,0,0,0,0,0,0,0,0]},\\"b\\":0}}"`;

exports[`DataView > toCrossJSONStream > supports DataView 1`] = `"{\\"t\\":28,\\"i\\":0,\\"f\\":{\\"t\\":41,\\"i\\":1,\\"s\\":2}}"`;

exports[`DataView > toCrossJSONStream > supports DataView 2`] = `"{\\"t\\":29,\\"i\\":0,\\"a\\":[{\\"t\\":41,\\"i\\":4,\\"s\\":3},{\\"t\\":22,\\"i\\":2,\\"l\\":16,\\"f\\":{\\"t\\":21,\\"i\\":3,\\"s\\":[0,0,42,0,0,0,0,0,0,0,0,0,0,0,0,0]},\\"b\\":0}]}"`;

exports[`DataView > toJSON > supports DataView 1`] = `"{\\"t\\":{\\"t\\":22,\\"i\\":0,\\"l\\":16,\\"f\\":{\\"t\\":21,\\"i\\":1,\\"s\\":[0,0,42,0,0,0,0,0,0,0,0,0,0,0,0,0]},\\"b\\":0},\\"f\\":16383,\\"m\\":[]}"`;

exports[`DataView > toJSONAsync > supports DataView 1`] = `"{\\"t\\":{\\"t\\":12,\\"i\\":0,\\"s\\":1,\\"f\\":{\\"t\\":22,\\"i\\":1,\\"l\\":16,\\"f\\":{\\"t\\":21,\\"i\\":2,\\"s\\":[0,0,42,0,0,0,0,0,0,0,0,0,0,0,0,0]},\\"b\\":0}},\\"f\\":16383,\\"m\\":[]}"`;
